File: code/ch08/8.5.1_1206.skip-list.java
class Skiplist {
  class SkipListNode {
      int val;
      int cnt;  // 当前val出现的次数
      SkipListNode[] levels;  // start from 0
      SkipListNode() {
          levels = new SkipListNode[MAX_LEVEL];
      }
  }

  private double p = 0.5;
  private int MAX_LEVEL = 16;
  private SkipListNode head;  // 头结点
  private int level;  // 
  private Random random;

  public Skiplist() {

      level = 0;  
      head = new SkipListNode();
      random = new Random();
  }

  public boolean search(int target) {
      SkipListNode curNode = head;
      for (int i = level-1; i >= 0; i--) {
          while (curNode.levels[i] != null && curNode.levels[i].val < target) {
              curNode = curNode.levels[i];
          }
      }
      curNode = curNode.levels[0];
      return (curNode != null && curNode.val == target);
  }

  public void add(int num) {
      SkipListNode curNode = head;
      // 记录每层能访问的最右节点
      SkipListNode[] levelTails = new SkipListNode[MAX_LEVEL];
      for (int i = level-1; i >= 0; i--) {
          while (curNode.levels[i] != null && curNode.levels[i].val < num) {
              curNode = curNode.levels[i];
          }
          levelTails[i] = curNode;
      }
      curNode = curNode.levels[0];
      if (curNode != null && curNode.val == num) {
          // 已存在，cnt 加1
          curNode.cnt++;
      } else {
          // 插入
          int newLevel = randomLevel();
          if (newLevel > level) {
              for (int i = level; i < newLevel; i++) {
                  levelTails[i] = head;
              }
              level = newLevel;
          }
          SkipListNode newNode = new SkipListNode();
          newNode.val = num;
          newNode.cnt = 1;
          for (int i = 0; i < level; i++) {
              newNode.levels[i] = levelTails[i].levels[i];
              levelTails[i].levels[i] = newNode;
              
          }
      }
  }

  private int randomLevel() {
      int level = 1;  
      while (random.nextDouble() < p && level < MAX_LEVEL) {
          level++;
      }
      return level > MAX_LEVEL ? MAX_LEVEL : level;
  }
  public boolean erase(int num) {
      SkipListNode curNode = head;
      // 记录每层能访问的最右节点
      SkipListNode[] levelTails = new SkipListNode[MAX_LEVEL];

      for (int i = level-1; i >= 0; i--) {
          while (curNode.levels[i] != null && curNode.levels[i].val < num) {
              curNode = curNode.levels[i];
          }
          levelTails[i] = curNode;
      }
      curNode = curNode.levels[0];
      if (curNode != null && curNode.val == num) {
          if (curNode.cnt > 1) {
              curNode.cnt--;
              return true;
          }
          for (int i = 0; i < level; i++) {
              if (levelTails[i].levels[i] != curNode) {
                  break;
              }
              levelTails[i].levels[i] = curNode.levels[i];
          }
          while (level > 0 && head.levels[level-1] == null) {
              level--;
          }
          return true;
      } 
      return false;
  }
}

File: code/ch08/8.5.1_1206.skip-list.py
from typing import Optional
import random


class ListNode:
    def __init__(self, data: Optional[int] = None):
        self._data = data
        self._forwards = []


class Skiplist:

    _MAX_LEVEL = 16

    def __init__(self):
        self._level_count = 1
        self._head = ListNode()
        self._head._forwards = [None] * self._MAX_LEVEL

    def search(self, target: int) -> bool:
        p = self._head
        for i in range(self._level_count - 1, -1, -1):
            while p._forwards[i] and p._forwards[i]._data < target:
                p = p._forwards[i]

        if p._forwards[0] and p._forwards[0]._data == target:
            return True

        return False

    def add(self, num: int) -> None:
        level = self._random_level()
        if self._level_count < level:
            self._level_count = level
        new_node = ListNode(num)
        new_node._forwards = [None] * level
        update = [self._head] * self._level_count

        p = self._head
        for i in range(self._level_count - 1, -1, -1):
            while p._forwards[i] and p._forwards[i]._data < num:
                p = p._forwards[i]

            update[i] = p

        for i in range(level):
            new_node._forwards[i] = update[i]._forwards[i]
            update[i]._forwards[i] = new_node

    def erase(self, num: int) -> bool:
        update = [None] * self._level_count
        p = self._head
        for i in range(self._level_count - 1, -1, -1):
            while p._forwards[i] and p._forwards[i]._data < num:
                p = p._forwards[i]
            update[i] = p

        if p._forwards[0] and p._forwards[0]._data == num:
            for i in range(self._level_count - 1, -1, -1):
                if update[i]._forwards[i] and update[i]._forwards[i]._data == num:
                    update[i]._forwards[i] = update[i]._forwards[i]._forwards[i]
            return True

        while self._level_count > 1 and not self._head._forwards[self._level_count]:
            self._level_count -= 1

        return False

    def _random_level(self, p: float = 0.5) -> int:
        level = 1
        while random.random() < p and level < self._MAX_LEVEL:
            level += 1
        return level


# Your Skiplist object will be instantiated and called as such:
# obj = Skiplist()
# param_1 = obj.search(target)
# obj.add(num)
# param_3 = obj.erase(num)
